File: moves.go
package solver

import (
	"fmt"
	"strings"
)

//          00 08 01
//          09 20 10
//          02 11 03
// 00 09 02 02 11 03 03 10 02 02 08 00
// 18 24 12 12 21 13 13 25 19 19 23 18
// 06 15 04 04 14 05 05 16 07 07 17 06
//          04 14 05
//          15 22 16
//          06 17 07

const U1Num, U2Num, U3Num byte = 0, 1, 2
const F1Num, F2Num, F3Num byte = 3, 4, 5
const D1Num, D2Num, D3Num byte = 6, 7, 8
const B1Num, B2Num, B3Num byte = 9, 10, 11
const L1Num, L2Num, L3Num byte = 12, 13, 14
const R1Num, R2Num, R3Num byte = 15, 16, 17
const M1Num, M2Num, M3Num byte = 18, 19, 20
const E1Num, E2Num, E3Num byte = 21, 22, 23
const S1Num, S2Num, S3Num byte = 24, 25, 26

var moveNames = []string{
	"U", "U2", "U'",
	"F", "F2", "F'",
	"D", "D2", "D'",
	"B", "B2", "B'",
	"L", "L2", "L'",
	"R", "R2", "R'",
	"M", "M2", "M'",
	"E", "E2", "E'",
	"S", "S2", "S'",
}

var moveFuncs = []func(*Cube){
	U1, U2, U3,
	F1, F2, F3,
	D1, D2, D3,
	B1, B2, B3,
	L1, L2, L3,
	R1, R2, R3,
	M1, M2, M3,
	E1, E2, E3,
	S1, S2, S3,
}

func invertMove(move byte) byte {
	switch move % 3 {
	case 0:
		return move + 2
	case 1:
		return move
	case 2:
		return move - 2
	}
	panic("unreachable")
}

func InvertAlgorithm(alg Algorithm) Algorithm {
	out := make(Algorithm, len(alg))
	for i, move := range alg {
		out[len(alg)-i-1] = invertMove(move)
	}
	return out
}

const noCancel = 0xFF
const perfectCancel = 0xFE

var equivalences = [][]byte{
	{U2Num, U3Num, perfectCancel},
	{U3Num, perfectCancel, U1Num},
	{perfectCancel, U1Num, U2Num},
	{F2Num, F3Num, perfectCancel},
	{F3Num, perfectCancel, F1Num},
	{perfectCancel, F1Num, F2Num},
	{D2Num, D3Num, perfectCancel},
	{D3Num, perfectCancel, D1Num},
	{perfectCancel, D1Num, D2Num},
	{B2Num, B3Num, perfectCancel},
	{B3Num, perfectCancel, B1Num},
	{perfectCancel, B1Num, B2Num},
	{L2Num, L3Num, perfectCancel},
	{L3Num, perfectCancel, L1Num},
	{perfectCancel, L1Num, L2Num},
	{R2Num, R3Num, perfectCancel},
	{R3Num, perfectCancel, R1Num},
	{perfectCancel, R1Num, R2Num},
}

// Returns noCancel if the moves cannot be canceled.
// Returns perfectCancel if the moves perfectly cancel.
// Returns the move that remains if the moves can be simplified.
// For example, U1 and U2 can simplify to U3.
func cancelPairOfMoves(m1, m2 byte) byte {
	face1 := m1 / 3
	face2 := m2 / 3
	if face1 != face2 {
		return noCancel
	}
	return equivalences[m1][m2%3]
}

func algString(forward Algorithm, inverse Algorithm) string {
	inverted := InvertAlgorithm(inverse)
	combined := appendImmutable(forward, inverted...)
	cleaned := make(Algorithm, 0, len(combined))
	for _, move := range combined {
		if len(cleaned) > 0 {
			cancel := cancelPairOfMoves(cleaned[len(cleaned)-1], move)
			if cancel == noCancel {
				cleaned = append(cleaned, move)
			} else if cancel == perfectCancel {
				cleaned = cleaned[:len(cleaned)-1]
			} else {
				cleaned[len(cleaned)-1] = cancel
			}
		} else {
			cleaned = append(cleaned, move)
		}
	}
	movesAsStrings := make([]string, len(cleaned))
	for i, m := range cleaned {
		movesAsStrings[i] = moveNames[m]
	}
	return strings.Join(movesAsStrings, " ")
}

func StringToAlg(algStr string) Algorithm {
	moveStrings := strings.Split(algStr, " ")
	moves := make(Algorithm, 0, len(moveStrings))
	for _, move := range moveStrings {
		switch move {
		case "U":
			moves = append(moves, U1Num)
		case "U2":
			moves = append(moves, U2Num)
		case "U'":
			moves = append(moves, U3Num)
		case "F":
			moves = append(moves, F1Num)
		case "F2":
			moves = append(moves, F2Num)
		case "F'":
			moves = append(moves, F3Num)
		case "D":
			moves = append(moves, D1Num)
		case "D2":
			moves = append(moves, D2Num)
		case "D'":
			moves = append(moves, D3Num)
		case "B":
			moves = append(moves, B1Num)
		case "B2":
			moves = append(moves, B2Num)
		case "B'":
			moves = append(moves, B3Num)
		case "L":
			moves = append(moves, L1Num)
		case "L2":
			moves = append(moves, L2Num)
		case "L'":
			moves = append(moves, L3Num)
		case "R":
			moves = append(moves, R1Num)
		case "R2":
			moves = append(moves, R2Num)
		case "R'":
			moves = append(moves, R3Num)
		case "M":
			moves = append(moves, M1Num)
		case "M2":
			moves = append(moves, M2Num)
		case "M'":
			moves = append(moves, M3Num)
		case "E":
			moves = append(moves, E1Num)
		case "E2":
			moves = append(moves, E2Num)
		case "E'":
			moves = append(moves, E3Num)
		case "S":
			moves = append(moves, S1Num)
		case "S2":
			moves = append(moves, S2Num)
		case "S'":
			moves = append(moves, S3Num)
		case "x":
			moves = append(moves, R1Num, M3Num, L3Num)
		case "x2":
			moves = append(moves, R2Num, M2Num, L2Num)
		case "x'":
			moves = append(moves, R3Num, M1Num, L1Num)
		case "y":
			moves = append(moves, U1Num, E3Num, D3Num)
		case "y2":
			moves = append(moves, U2Num, E2Num, D2Num)
		case "y'":
			moves = append(moves, U3Num, E1Num, D1Num)
		case "z":
			moves = append(moves, F1Num, S1Num, B3Num)
		case "z2":
			moves = append(moves, F2Num, S2Num, B2Num)
		case "z'":
			moves = append(moves, F3Num, S3Num, B1Num)
		case "l":
			moves = append(moves, L1Num, M1Num)
		case "l2":
			moves = append(moves, L2Num, M2Num)
		case "l'":
			moves = append(moves, L3Num, M3Num)
		case "r":
			moves = append(moves, R1Num, M3Num)
		case "r2":
			moves = append(moves, R2Num, M2Num)
		case "r'":
			moves = append(moves, R3Num, M1Num)
		default:
			fmt.Printf("Unknown move: %s\n", move)
		}
	}
	return moves
}

func PerformAlgorithm(cube *Cube, alg Algorithm) {
	for _, move := range alg {
		moveFunc := moveFuncs[move]
		moveFunc(cube)
	}
}

func PerformAlgString(cube *Cube, algStr string) {
	alg := StringToAlg(algStr)
	PerformAlgorithm(cube, alg)
}

// Twist a corner clockwise
func twistCW(b byte) byte {
	if b == disregard {
		return b
	}
	upper := b & 0xF0
	lower := b & 0x0F
	return ((upper+1)%3)<<4 | lower
}

// Twist a corner counter-clockwise
func twistCCW(b byte) byte {
	if b == disregard {
		return b
	}
	upper := b & 0xF0
	lower := b & 0x0F
	return ((upper+2)%3)<<4 | lower

}

// Flip an edge
func flip(b byte) byte {
	if b == disregard {
		return b
	}
	return b ^ 0b00010000
}

func U1(c *Cube) {
	temp := c[0]
	c[0] = c[2]
	c[2] = c[3]
	c[3] = c[1]
	c[1] = temp

	temp = c[8]
	c[8] = c[9]
	c[9] = c[11]
	c[11] = c[10]
	c[10] = temp
}

func U2(c *Cube) {
	temp := c[0]
	c[0] = c[3]
	c[3] = temp

	temp = c[1]
	c[1] = c[2]
	c[2] = temp

	temp = c[8]
	c[8] = c[11]
	c[11] = temp

	temp = c[9]
	c[9] = c[10]
	c[10] = temp
}

func U3(c *Cube) {
	temp := c[0]
	c[0] = c[1]
	c[1] = c[3]
	c[3] = c[2]
	c[2] = temp

	temp = c[8]
	c[8] = c[10]
	c[10] = c[11]
	c[11] = c[9]
	c[9] = temp
}

func F1(c *Cube) {
	temp := c[2]
	c[2] = twistCCW(c[4])
	c[4] = twistCW(c[5])
	c[5] = twistCCW(c[3])
	c[3] = twistCW(temp)

	temp = c[11]
	c[11] = flip(c[12])
	c[12] = flip(c[14])
	c[14] = flip(c[13])
	c[13] = flip(temp)
}

func F2(c *Cube) {
	temp := c[2]
	c[2] = c[5]
	c[5] = temp

	temp = c[3]
	c[3] = c[4]
	c[4] = temp

	temp = c[11]
	c[11] = c[14]
	c[14] = temp

	temp = c[12]
	c[12] = c[13]
	c[13] = temp
}

func F3(c *Cube) {
	temp := c[2]
	c[2] = twistCCW(c[3])
	c[3] = twistCW(c[5])
	c[5] = twistCCW(c[4])
	c[4] = twistCW(temp)

	temp = c[11]
	c[11] = flip(c[13])
	c[13] = flip(c[14])
	c[14] = flip(c[12])
	c[12] = flip(temp)
}

func D1(c *Cube) {
	temp := c[4]
	c[4] = c[6]
	c[6] = c[7]
	c[7] = c[5]
	c[5] = temp

	temp = c[14]
	c[14] = c[15]
	c[15] = c[17]
	c[17] = c[16]
	c[16] = temp
}

func D2(c *Cube) {
	temp := c[4]
	c[4] = c[7]
	c[7] = temp

	temp = c[5]
	c[5] = c[6]
	c[6] = temp

	temp = c[14]
	c[14] = c[17]
	c[17] = temp

	temp = c[15]
	c[15] = c[16]
	c[16] = temp
}

func D3(c *Cube) {
	temp := c[4]
	c[4] = c[5]
	c[5] = c[7]
	c[7] = c[6]
	c[6] = temp

	temp = c[14]
	c[14] = c[16]
	c[16] = c[17]
	c[17] = c[15]
	c[15] = temp
}

func B1(c *Cube) {
	temp := c[0]
	c[0] = twistCW(c[1])
	c[1] = twistCCW(c[7])
	c[7] = twistCW(c[6])
	c[6] = twistCCW(temp)

	temp = c[8]
	c[8] = flip(c[19])
	c[19] = flip(c[17])
	c[17] = flip(c[18])
	c[18] = flip(temp)
}

func B2(c *Cube) {
	temp := c[0]
	c[0] = c[7]
	c[7] = temp

	temp = c[1]
	c[1] = c[6]
	c[6] = temp

	temp = c[8]
	c[8] = c[17]
	c[17] = temp

	temp = c[19]
	c[19] = c[18]
	c[18] = temp
}

func B3(c *Cube) {
	temp := c[0]
	c[0] = twistCW(c[6])
	c[6] = twistCCW(c[7])
	c[7] = twistCW(c[1])
	c[1] = twistCCW(temp)

	temp = c[8]
	c[8] = flip(c[18])
	c[18] = flip(c[17])
	c[17] = flip(c[19])
	c[19] = flip(temp)
}

func L1(c *Cube) {
	temp := c[0]
	c[0] = twistCCW(c[6])
	c[6] = twistCW(c[4])
	c[4] = twistCCW(c[2])
	c[2] = twistCW(temp)

	temp = c[9]
	c[9] = c[18]
	c[18] = c[15]
	c[15] = c[12]
	c[12] = temp
}

func L2(c *Cube) {
	temp := c[0]
	c[0] = c[4]
	c[4] = temp

	temp = c[2]
	c[2] = c[6]
	c[6] = temp

	temp = c[9]
	c[9] = c[15]
	c[15] = temp

	temp = c[12]
	c[12] = c[18]
	c[18] = temp
}

func L3(c *Cube) {
	temp := c[0]
	c[0] = twistCCW(c[2])
	c[2] = twistCW(c[4])
	c[4] = twistCCW(c[6])
	c[6] = twistCW(temp)

	temp = c[9]
	c[9] = c[12]
	c[12] = c[15]
	c[15] = c[18]
	c[18] = temp
}

func R1(c *Cube) {
	temp := c[1]
	c[1] = twistCW(c[3])
	c[3] = twistCCW(c[5])
	c[5] = twistCW(c[7])
	c[7] = twistCCW(temp)

	temp = c[10]
	c[10] = c[13]
	c[13] = c[16]
	c[16] = c[19]
	c[19] = temp
}

func R2(c *Cube) {
	temp := c[1]
	c[1] = c[5]
	c[5] = temp

	temp = c[3]
	c[3] = c[7]
	c[7] = temp

	temp = c[10]
	c[10] = c[16]
	c[16] = temp

	temp = c[13]
	c[13] = c[19]
	c[19] = temp
}

func R3(c *Cube) {
	temp := c[1]
	c[1] = twistCW(c[7])
	c[7] = twistCCW(c[5])
	c[5] = twistCW(c[3])
	c[3] = twistCCW(temp)

	temp = c[10]
	c[10] = c[19]
	c[19] = c[16]
	c[16] = c[13]
	c[13] = temp
}

func M1(c *Cube) {
	temp := c[8]
	c[8] = flip(c[17])
	c[17] = flip(c[14])
	c[14] = flip(c[11])
	c[11] = flip(temp)

	temp = c[20]
	c[20] = c[23]
	c[23] = c[22]
	c[22] = c[21]
	c[21] = temp
}

func M2(c *Cube) {
	temp := c[8]
	c[8] = c[14]
	c[14] = temp

	temp = c[11]
	c[11] = c[17]
	c[17] = temp

	temp = c[20]
	c[20] = c[22]
	c[22] = temp

	temp = c[21]
	c[21] = c[23]
	c[23] = temp
}

func M3(c *Cube) {
	temp := c[8]
	c[8] = flip(c[11])
	c[11] = flip(c[14])
	c[14] = flip(c[17])
	c[17] = flip(temp)

	temp = c[20]
	c[20] = c[21]
	c[21] = c[22]
	c[22] = c[23]
	c[23] = temp
}

func E1(c *Cube) {
	temp := c[12]
	c[12] = flip(c[18])
	c[18] = flip(c[19])
	c[19] = flip(c[13])
	c[13] = flip(temp)

	temp = c[21]
	c[21] = c[24]
	c[24] = c[23]
	c[23] = c[25]
	c[25] = temp
}

func E2(c *Cube) {
	temp := c[12]
	c[12] = c[19]
	c[19] = temp

	temp = c[13]
	c[13] = c[18]
	c[18] = temp

	temp = c[21]
	c[21] = c[23]
	c[23] = temp

	temp = c[24]
	c[24] = c[25]
	c[25] = temp
}

func E3(c *Cube) {
	temp := c[12]
	c[12] = flip(c[13])
	c[13] = flip(c[19])
	c[19] = flip(c[18])
	c[18] = flip(temp)

	temp = c[21]
	c[21] = c[25]
	c[25] = c[23]
	c[23] = c[24]
	c[24] = temp
}

func S1(c *Cube) {
	temp := c[9]
	c[9] = flip(c[15])
	c[15] = flip(c[16])
	c[16] = flip(c[10])
	c[10] = flip(temp)

	temp = c[20]
	c[20] = c[24]
	c[24] = c[22]
	c[22] = c[25]
	c[25] = temp
}

func S2(c *Cube) {
	temp := c[9]
	c[9] = c[16]
	c[16] = temp

	temp = c[10]
	c[10] = c[15]
	c[15] = temp

	temp = c[20]
	c[20] = c[22]
	c[22] = temp

	temp = c[24]
	c[24] = c[25]
	c[25] = temp
}

func S3(c *Cube) {
	temp := c[9]
	c[9] = flip(c[10])
	c[10] = flip(c[16])
	c[16] = flip(c[15])
	c[15] = flip(temp)

	temp = c[20]
	c[20] = c[25]
	c[25] = c[22]
	c[22] = c[24]
	c[24] = temp
}

File: solver/src/moves.rs
use arrayvec::ArrayVec;

use crate::algorithm::{Algorithm, AlgorithmSegment};

pub const U1_NUM: u8 = 0;
pub const U2_NUM: u8 = 1;
pub const U3_NUM: u8 = 2;
pub const F1_NUM: u8 = 3;
pub const F2_NUM: u8 = 4;
pub const F3_NUM: u8 = 5;
pub const D1_NUM: u8 = 6;
pub const D2_NUM: u8 = 7;
pub const D3_NUM: u8 = 8;
pub const B1_NUM: u8 = 9;
pub const B2_NUM: u8 = 10;
pub const B3_NUM: u8 = 11;
pub const L1_NUM: u8 = 12;
pub const L2_NUM: u8 = 13;
pub const L3_NUM: u8 = 14;
pub const R1_NUM: u8 = 15;
pub const R2_NUM: u8 = 16;
pub const R3_NUM: u8 = 17;
pub const M1_NUM: u8 = 18;
pub const M2_NUM: u8 = 19;
pub const M3_NUM: u8 = 20;
pub const E1_NUM: u8 = 21;
pub const E2_NUM: u8 = 22;
pub const E3_NUM: u8 = 23;
pub const S1_NUM: u8 = 24;
pub const S2_NUM: u8 = 25;
pub const S3_NUM: u8 = 26;
pub const NULL_MOVE: u8 = 0xFF;

pub const MOVE_NAMES: [&str; 27] = [
    "U", "U2", "U'",
    "F", "F2", "F'",
    "D", "D2", "D'",
    "B", "B2", "B'",
    "L", "L2", "L'",
    "R", "R2", "R'",
    "M", "M2", "M'",
    "E", "E2", "E'",
    "S", "S2", "S'",
];

const NO_CANCEL: u8 = 0xFF;
const PERFECT_CANCEL: u8 = 0xFE;

const EQUIVALENCES: [[u8; 3]; 27] = [
    [U2_NUM, U3_NUM, PERFECT_CANCEL],
    [U3_NUM, PERFECT_CANCEL, U1_NUM],
    [PERFECT_CANCEL, U1_NUM, U2_NUM],
    [F2_NUM, F3_NUM, PERFECT_CANCEL],
    [F3_NUM, PERFECT_CANCEL, F1_NUM],
    [PERFECT_CANCEL, F1_NUM, F2_NUM],
    [D2_NUM, D3_NUM, PERFECT_CANCEL],
    [D3_NUM, PERFECT_CANCEL, D1_NUM],
    [PERFECT_CANCEL, D1_NUM, D2_NUM],
    [B2_NUM, B3_NUM, PERFECT_CANCEL],
    [B3_NUM, PERFECT_CANCEL, B1_NUM],
    [PERFECT_CANCEL, B1_NUM, B2_NUM],
    [L2_NUM, L3_NUM, PERFECT_CANCEL],
    [L3_NUM, PERFECT_CANCEL, L1_NUM],
    [PERFECT_CANCEL, L1_NUM, L2_NUM],
    [R2_NUM, R3_NUM, PERFECT_CANCEL],
    [R3_NUM, PERFECT_CANCEL, R1_NUM],
    [PERFECT_CANCEL, R1_NUM, R2_NUM],
    [M2_NUM, M3_NUM, PERFECT_CANCEL],
    [M3_NUM, PERFECT_CANCEL, M1_NUM],
    [PERFECT_CANCEL, M1_NUM, M2_NUM],
    [E2_NUM, E3_NUM, PERFECT_CANCEL],
    [E3_NUM, PERFECT_CANCEL, E1_NUM],
    [PERFECT_CANCEL, E1_NUM, E2_NUM],
    [S2_NUM, S3_NUM, PERFECT_CANCEL],
    [S3_NUM, PERFECT_CANCEL, S1_NUM],
    [PERFECT_CANCEL, S1_NUM, S2_NUM],
];

pub fn invert_move(m: u8) -> u8 {
    match m % 3 {
        0 => m + 2,
        1 => m,
        2 => m - 2,
        _ => panic!("unreachable"),
    }
}

pub fn invert_algorithm(alg: Algorithm) -> Algorithm {
    let mut out = Algorithm::new();
    for m in alg {
        out.push(invert_move(m));
    }
    out.reverse();
    out
}

pub fn invert_alg_segment(alg: AlgorithmSegment) -> AlgorithmSegment {
    let mut out = ArrayVec::new();
    for m in alg {
        out.push(invert_move(m));
    }
    out.reverse();
    out
}

fn cancel_pair_of_moves(m1: u8, m2: u8) -> u8 {
    let face1 = m1 / 3;
    let face2 = m2 / 3;
    if face1 != face2 {
        return NO_CANCEL;
    }
    EQUIVALENCES[m1 as usize][m2 as usize % 3]
}

pub fn combine_algs(forward: AlgorithmSegment, inverse: AlgorithmSegment) -> String {
    let mut combined = Algorithm::new();
    combined.extend(forward);
    combined.extend(inverse);
    let mut cleaned: Vec<u8> = Vec::new();
    for m in combined {
        let len = cleaned.len();
        if len > 0 {
            let cancel = cancel_pair_of_moves(cleaned[cleaned.len() - 1], m);
            if cancel == NO_CANCEL {
                cleaned.push(m);
            } else if cancel == PERFECT_CANCEL {
                cleaned.pop();
            } else {
                cleaned[len - 1] = cancel;
            }
        } else {
            cleaned.push(m);
        }
    }
    cleaned
        .iter()
        .map(|m| MOVE_NAMES[*m as usize])
        .collect::<Vec<&str>>()
        .join(" ")
}

// The moves that can be performed while searching for a solution.
// For example, the user may want to restrict solutions to the RUF moves.
pub struct Moves {
    moves: Vec<u8>,
}

impl Moves {

    // This function accepts a string of comma-separated move names.
    // Example: "U,U2,U',F,F2,F',R,R2,R'"
    pub fn from_string(s: &str) -> Moves {
        let mut moves = Vec::new();
        for m in s.split(' ') {
            match m {
                "U" => moves.push(U1_NUM),
                "U2" => moves.push(U2_NUM),
                "U'" => moves.push(U3_NUM),
                "F" => moves.push(F1_NUM),
                "F2" => moves.push(F2_NUM),
                "F'" => moves.push(F3_NUM),
                "D" => moves.push(D1_NUM),
                "D2" => moves.push(D2_NUM),
                "D'" => moves.push(D3_NUM),
                "B" => moves.push(B1_NUM),
                "B2" => moves.push(B2_NUM),
                "B'" => moves.push(B3_NUM),
                "L" => moves.push(L1_NUM),
                "L2" => moves.push(L2_NUM),
                "L'" => moves.push(L3_NUM),
                "R" => moves.push(R1_NUM),
                "R2" => moves.push(R2_NUM),
                "R'" => moves.push(R3_NUM),
                _ => panic!("invalid move: {}", m),
            }
        }
        Moves { moves }
    }

    pub fn get_moves(&self) -> &Vec<u8> {
        &self.moves
    }

    // Returns true if the moves contain one or more double moves.
    pub fn has_double(&self) -> bool {
        for m in &self.moves {
            if m % 3 == 1 {
                return true;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use crate::moves::*;

    #[test]
    fn test_moves() {
        let moves = Moves::from_string("U U2 U' F F2 F' R R2 R'");
        assert_eq!(*moves.get_moves(), vec![U1_NUM, U2_NUM, U3_NUM, F1_NUM, F2_NUM, F3_NUM, R1_NUM, R2_NUM, R3_NUM]);
    }
}
